src/main/java/org/xbill/DNS/Client.java

```diff
@@ -17,39 +18,47 @@
   /** Packet logger, if available. */
   private static PacketLogger packetLogger = null;
 
-  private static volatile boolean run;
+  private static volatile boolean run = true;
+  private static final List<Runnable> timeoutTasks = new CopyOnWriteArrayList<>();
+  private static final List<Runnable> closeTasks = new CopyOnWriteArrayList<>();
   private static Thread selectorThread;
-  private static List<Runnable> timeoutTasks = new CopyOnWriteArrayList<>();
-  static Selector selector;
+  private static volatile Selector selector;
 
   interface KeyProcessor {
     void processReadyKey(SelectionKey key);
   }
 
-  static void start() throws IOException {
-    if (run) {
-      return;
+  static Selector selector() throws IOException {
+    if (selector == null) {
+      synchronized (Client.class) {
+        if (selector == null) {
+          selector = Selector.open();
+          log.debug("Starting dnsjava NIO selector thread");
+          selectorThread = new Thread(Client::runSelector);
+          selectorThread.setDaemon(true);
+          selectorThread.setName("dnsjava NIO selector");
+          selectorThread.start();
+          Thread closeThread = new Thread(Client::close);
+          closeThread.setName("dnsjava NIO shutdown hook");
+          Runtime.getRuntime().addShutdownHook(closeThread);
+        }
+      }
     }
 
-    log.debug("Starting dnsjava NIO selector thread");
-    run = true;
-    selector = Selector.open();
-    selectorThread = new Thread(Client::runSelector);
-    selectorThread.setDaemon(true);
-    selectorThread.setName("dnsjava NIO selector");
-    selectorThread.start();
+    return selector;
   }
 
-  protected static void close() throws Exception {
-    if (!run) {
-      return;
-    }
-
+  private static void close() {
     run = false;
+    closeTasks.forEach(Runnable::run);
     timeoutTasks.clear();
     selector.wakeup();
-    selector.close();
-    selectorThread.join();
+    try {
+      selector.close();
+      selectorThread.join();
```

Comment:
> If `selector.close()` returns an `IOException` (as its documentation says it could do), `selectorThread.join()` will not execute. It may be more ideal to use try-finally here to ensure both operations run.

This is just intended to be nice during a normal all-is-nice shutdown. `close` is _only_ called when the program ends and is not guaranteed to run at all. If closing the selector fails, okay, no matter. It is a daemon thread not holding any resources which will end anyway.

> 
> In my code, I am typically very pedantic about close methods and use try-finally nested blocks around each statement that could possibly throw an exception, regardless of how rare the exception may be. Ie. something like,
> 
> ```
>   private static void close() {
>     run = false;
>     initialisationCompleted.await();

The shutdown hook definitely should not wait for an initialization. But it is pointless anyway, as the initialization is definitely complete by the time this can be called (by the shutdown hook).

>     try {
>       try {
>         for(Runnable nextCloseTask : closeTasks) {
>           try {
>             nextCloseTask.run();
>           } catch (Exception closeTaskException) {
>             log.warn("Close task failed: ", closeTaskException);

The only possible exceptions to catch here are `RuntimeException`s, which I wouldn't really want to catch.

>           }
>         }
>       } finally {
>         try {
>           timeoutTasks.clear();
>         } finally {
>           try {
>             Selector selectorToClose = selector;
>             if(selectorToClose != null) {

This is impossible. The selector is never null by the time `close` is executed (unless `Selector.open()` would return null, but then everything is broken).

>               selectorToClose.wakeup();
>               selectorToClose.close();
>             }
>           } finally {
>             Thread selectorThreadToJoin = selectorThread;
>             if(selectorThreadToJoin != null) {

This is impossible. The selectorThread is never null by the time `close` is executed.

>               selectorThreadToJoin.join();
>             }
>           }
>         }
>     } catch (InterruptedException | IOException e) {
>       log.warn("Failed to properly shutdown", e);
>     }
> }
> ```